File: src/main/java/org/blacksun/graph/channel/DuplexChannel.java
package org.blacksun.graph.channel;

import org.blacksun.graph.node.GraphNode;

public class DuplexChannel extends AbstractChannel {
    protected DuplexChannel pair;

    private DuplexChannel(DuplexChannel channel) {
        super(channel.toNode, channel.fromNode, channel.weight, channel.errors);
        pair = channel;
    }

    public DuplexChannel(GraphNode fromNode, GraphNode toNode, int weight, double errors) {
        super(fromNode, toNode, weight, errors);
        pair = new DuplexChannel(this);
    }

    @Override
    public void connect() {
        super.connect();
        toNode.addConnection(pair);
    }

    @Override
    public void remove() {
        super.remove();
        toNode.removeConnection(pair);
    }

    @Override
    protected String getDirection() {
        return "<==>";
    }
}

File: src/main/java/org/blacksun/graph/channel/DuplexChannel.kt
package org.blacksun.graph.channel

import org.blacksun.graph.node.GraphNode

class DuplexChannel : AbstractChannel {
    private var pair: DuplexChannel

    override val direction = "<==>"

    private constructor(channel: DuplexChannel) : super(channel.toNode, channel.fromNode, channel.weight, channel.errors) {
        pair = channel
    }

    constructor(fromNode: GraphNode, toNode: GraphNode, weight: Int, errors: Double) : super(fromNode, toNode, weight, errors) {
        pair = DuplexChannel(this)
        connect()
    }

    override fun connect() {
        super.connect()
        toNode.addConnection(pair)
    }

    override fun remove() {
        super.remove()
        toNode.removeConnection(pair)
    }
}
